applications/FluidDynamicsApplication/custom_utilities/mass_conservation_utility.cpp

```diff
@@ -30,131 +30,195 @@ namespace Kratos
-    output_line_timestep +=             std::to_string(water_volume_error) + "\t\t";
+    output_line_timestep +=             std::to_string(mWaterVolumeError) + "\t\t";
     output_line_timestep +=             std::to_string(net_inflow_inlet) + "\t\t";
     output_line_timestep +=             std::to_string(net_inflow_outlet) + "\t\t";
-    output_line_timestep +=             std::to_string(inter_area) + "\t\t";
-    output_line_timestep +=             std::to_string(shift_for_correction) +"\n";
+    output_line_timestep +=             std::to_string(inter_area) + "\n";
+
     return output_line_timestep;
 }
 
 
 
-void MassConservationCheckProcess::ComputeVolumesAndInterface( double& positiveVolume, double& negativeVolume, double& interfaceArea ){
+double MassConservationUtility::ComputeDtForConvection(){
+
+    // a small step is set to avoid numerical problems
+    double time_step_for_convection = 1.0e-7;
+    const auto& r_comm = mrModelPart.GetCommunicator();
+
+    if ( mWaterVolumeError > 0.0 ){
+        // case: water volume was lost by mistake
+        mAddWater = true;
+        double water_outflow_over_boundary = OrthogonalFlowIntoAir( 1.0 );
+        if ( r_comm.GetDataCommunicator().SumAll( water_outflow_over_boundary ) ){
+            // checking if flow is sufficient (avoid division by 0)
+            if ( water_outflow_over_boundary > 1.0e-12 ){
+                time_step_for_convection = mWaterVolumeError / water_outflow_over_boundary;
+            }
+        } else {
+            KRATOS_ERROR << "Communication failed in MassConservationUtility::ComputeDtForConvection()";
+        }
+    }
+    else if ( mWaterVolumeError < 0.0 && mCorrectBackwards) {
+        // case: water volume was gained by mistake
+        mAddWater = false;
+        double water_inflow_over_boundary = OrthogonalFlowIntoAir( -1.0 );
+        if ( r_comm.GetDataCommunicator().SumAll( water_inflow_over_boundary ) ){
+            // checking if flow is sufficient (avoid division by 0)
+            if ( water_inflow_over_boundary > 1.0e-12 ){
+                time_step_for_convection = - mWaterVolumeError / water_inflow_over_boundary;
+            }
+        } else {
+            KRATOS_ERROR << "Communication failed in MassConservationUtility::ComputeDtForConvection()";
+        }
+    }
+    else {
+        // case: Exactly the correct volume of water is present
+        mAddWater = true;
+    }
+
+    KRATOS_WARNING_IF("MassConservationUtility", time_step_for_convection < 0.0) << "A time step smaller than 0.0 was computed." << std::endl;
+
+    return time_step_for_convection;
+}
+
+
+
+void MassConservationUtility::ApplyLocalCorrection( Variable<double>& rAuxDistVar ){
+
+    const auto node_begin = mrModelPart.GetCommunicator().LocalMesh().NodesBegin();
+    const int number_nodes = static_cast<int>(mrModelPart.GetCommunicator().LocalMesh().NumberOfNodes());
+
+    for (int i_node = 0; i_node < number_nodes; ++i_node){
```

Comment:
```suggestion
#pragma omp parallel for firstprivate(node_begin)
    for (int i_node = 0; i_node < number_nodes; ++i_node){
```
I think this loop can be parallel.

Reply:
done, why do you  need the firstprivate(node_begin)? Performance?